connect-button/src/main/java/com/ifttt/connect/ui/AnalyticsManager.java

```diff
@@ -0,0 +1,257 @@
+         * Use the in-memory queue as fallback
+         **/
+        Moshi moshi = new Moshi.Builder().build();
+
+        try {
+            File folder = context.getDir("analytics-disk-queue", Context.MODE_PRIVATE);
+            QueueFile queueFile = createQueueFile(folder);
+            queue = ObjectQueue.create(queueFile, new AnalyticsPayloadConverter(moshi));
+        } catch (IOException e) {
+            queue = ObjectQueue.createInMemory();
+        }
+
+        workManager = WorkManager.getInstance(context);
+        preference = AnalyticsPreferences.getInstance(context);
+    }
+
+    static AnalyticsManager getInstance(Context context) {
+        if (INSTANCE == null) {
+            INSTANCE = new AnalyticsManager(context.getApplicationContext());
+        }
+        return INSTANCE;
+    }
+
+    /*
+     * This method generates a UI item impression event
+     * */
+    void trackUiImpression(AnalyticsObject obj, AnalyticsLocation location) {
+        trackItemEvent("android.impression", obj, location);
+    }
+
+    /*
+     * This method generates a UI item click event
+     * */
+    void trackUiClick(AnalyticsObject obj, AnalyticsLocation location) {
+        trackItemEvent("android.click", obj, location);
+    }
+
+    /*
+     * Process the event data before adding it to the event queue
+     * */
+    private void trackItemEvent(String name, AnalyticsObject obj, AnalyticsLocation location) {
+        if (preference.getAnalyticsTrackingOptOutPreference()) {
+            return;
+        }
+
+        Map<String, String> properties = new HashMap<>();
+
+        properties.put("object_id", obj.id);
+        properties.put("object_type", obj.type);
+        if (obj instanceof AnalyticsObject.ConnectionAnalyticsObject) {
+            properties.put("object_status", ((AnalyticsObject.ConnectionAnalyticsObject) obj).status);
+        }
+
+        properties.put("location_id", location.id);
+        properties.put("location_type", location.type);
+
+        properties.put("sdk_version", BuildConfig.VERSION_NAME);
+        properties.put("system_version", Integer.toString(VERSION.SDK_INT));
+        String timestamp = Long.toString(System.currentTimeMillis());
+
+        performEnqueue(new AnalyticsEventPayload(
+                name,
+                timestamp,
+                properties
+        ));
+    }
+
+    /*
+     * Returns the current size of the queue
+     * */
+    int getQueueSize() {
+        return queue.size();
+    }
+
+    /*
+     * Enqueue an async task to add to the queue.
+     * The default serial scheduler of AsyncTask will be used to ensure that events are enqueued in the correct order.
+     **/
+    private void performEnqueue(AnalyticsEventPayload payload) {
+        new EnqueueTask(this).execute(payload);
+    }
+
+    /*
+     * Schedules a one time work request to read from the queue, make an api call to submit events and remove them from queue
+     * */
+    static void performFlush() {
```

Comment:
If we find that we are losing some events with our current flush implementation, we may want to invoke this method from the `ConnectButton` after some pre-determined operation.

Reply:
I see. Since this method is relatively simple, I think we can add that if we need this use case. What do you think?

Also, it is only going to schedule a worker. I think it might make more sense to have a util method in the worker to schedule the work, rather than having this within the AnalyticsManager.

Reply:
yup, makes sense.